framework/src/base/NonlinearSystem.C

```diff
@@ -1786,16 +1795,65 @@ NonlinearSystem::computeJacobianInternal(SparseMatrix<Number> &  jacobian)
       {
         _fe_problem.reinitNodeFace(node, boundary_id, 0);
 
-        for (std::vector<NodalBC *>::const_iterator it = _bcs[0].getNodalBCs(boundary_id).begin(); it != _bcs[0].getNodalBCs(boundary_id).end(); ++it)
+        const std::vector<NodalBC *> & bcs = _bcs[0].getNodalBCs(boundary_id);
+        for (std::vector<NodalBC *>::const_iterator it = bcs.begin(); it != bcs.end(); ++it)
         {
           NodalBC * bc = *it;
-          if (bc->shouldApply() && bc->variable().isNodalDefined())
-            zero_rows.push_back(bc->variable().nodalDofIndex());
+
+          // Get the list of variables to which this BC is actually
+          // coupled.  We should not need to call
+          // computeOffDiagJacobian() for variables the BC is not
+          // coupled to.  Note: this list does not contain bc->variable.number().
+          const std::vector<MooseVariable *> & coupled_moose_vars = bc->getCoupledMooseVars();
+
+          // Loop over all the variables whose Jacobian blocks are
+          // actually being computed, call computeOffDiagJacobian()
+          // for each one which is actually coupled (otherwise the
+          // value is zero.)
+          for (std::vector<std::pair<MooseVariable *, MooseVariable *> >::iterator it = coupling_entries.begin();
+               it != coupling_entries.end(); ++it)
+          {
+            unsigned int
+              ivar = it->first->number(),
+              jvar = it->second->number();
+
+            // We are only going to do something if the BC's variable is the same as ivar and the BC should apply.
+            if ((bc->variable().number() == ivar) && bc->shouldApply())
+            {
+              // Is the BC actually coupled to jvar?
+              bool bc_is_coupled_to_jvar = false;
+              for (unsigned int var = 0; var < coupled_moose_vars.size(); ++var)
```

Comment:
Seems like we should process this first before we start the node loop and figure this out once - instead of doing it once per node.